File: Tree_traversal/traversal.cpp
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    ~TreeNode() {delete left, right, val;};
};

void BFS(TreeNode* root) {
    if (root == nullptr)
        return;
    vector<TreeNode*> level = {root};
    while (!level.empty()) {
        vector<TreeNode*> _temp;
        for (TreeNode* node: level) {
            if (node->left) 
                _temp.push_back(node->left);
            if (node->right)
                _temp.push_back(node->right);
        }
        swap(_temp, level);
    }
    return;
}

void DFS(TreeNode* root) {
    if (root == nullptr)
        return;
    DFS(root->left);
    // do something here
    DFS(root->right);
}

File: Tree_traversal/traversal.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def BFS(root : TreeNode):
    if not root:
        return -1
    dq = deque([root])
    while dq:
        node = dq.popleft()

        # do something here

        if node.left:
            dq.append(node.left)
        if node.right:
            dq.append(node.right)

def DFS(root : TreeNode):
    if not root:
        return -1
    
    s = [root]
    while s:

        while s[-1].left:
            s.append(s[-1].left)
        
        node = s.pop()
        # do something here
        if node.right:
            s.append(node.right)

def traversal(root: TreeNode):
    if not root:
        return []
    
    # preorder
    return [root.val] + traversal(root.left) + traversal(root.right)
    # inorder
    return  traversal(root.left) + [root.val] + traversal(root.right)
    # postorder
    return traversal(root.left) + traversal(root.right) + [root.val]